Threads API specification and implementation

kalvdans opened the pull request:
Fixes https://github.com/Couchers-org/couchers/issues/269

Initially only the API specification is committed, and later on when the implementation is done I will un-wip the PR.

Commit:
First draft of threads API specification

amdwit commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,105 @@
+  // Only get messages posted before the given comment id. Pass -1 to get all.
+  int64 posted_before_comment_id = 2;
+  // Maximum number of comments to return
+  uint32 number = 3;
```
The db and js cannot do this type exactly. The db can only do signed integers and js do float64 (and int32 in some circumstances).
```suggestion
  int32 number = 3;
```
Or maybe use int64 for all our integer use?

kalvdans replied:
This is the maximum length of the reply requested by client, it is never stored in the database. Javascript has 53-bit integers so should be able to handle uint32 just fine.

amdwit replied:
But isn't it better to always use int64 for our integers in proto, so we never have to think about it?

kalvdans replied:
no, int64 are converted to strings for Javascript, really hard to work with. For ID:s that's fine as the client should never has to do arithmetic on them. But for other usage like counting items, it is easier if we use integers of a size that is supported natively in Javascript.

We really need an int53 type...

amdwit replied:
> int64 are converted to strings

Are you sure about it? In the generated code you see `number` being used.

aapeliv replied:
We have used `uint32`s for counts elsewhere too.

amdwit replied:
But why not in those two cases below?

amdwit replied:
Wouldn't it be better to make it all int64? Given that `number` has been used in the generated (.ts) files for int64?

kalvdans replied:
To be honest, I haven't managed to generate the Javascript code :blush: But https://developers.google.com/protocol-buffers/docs/proto3#json says int64 and uint64 gets converted to string in JSON.

amdwit commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,105 @@
+
+  // number of comments posted earlier than the ones returned in `comments`
+  // zero if all comments has been returned to the client.
+  int64 nr_earlier_comments = 2;
```
This should also be uint32 then.
```suggestion
  uint32 nr_earlier_comments = 2;
```

amdwit commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,105 @@
+
+  // number of replies posted earlier than the ones returned in `replies`.
+  // zero if all replies has been returned to the client.
+  int64 nr_earlier_replies = 2;
```
```suggestion
  uint32 nr_earlier_replies = 2;
```

aapeliv commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,105 @@
+
+message GetCommentsReq {
+  int64 thread_id = 1;
+  // Only get messages posted before the given comment id. Pass -1 to get all.
+  int64 posted_before_comment_id = 2;
```
How about using 0 for unset? The IDs will come from postgres serials that start from 1, so IDs should never be 0?

lucaslcode replied:
Additional benefit: protobuf ints are 0 by default

aapeliv commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,105 @@
+  //    +- reply8
+
+  // Get non-deleted comments belonging to a thread
+  rpc GetComments(GetCommentsReq) returns (GetCommentsRes);
```
Could you please specify in a comment the ordering and logic of returned comments. It's not so clear at the moment. Something like

```py
// Returns the last `number` non-deleted comments before `posted_before_comment_id`
```

Am I understanding this correctly? So the client would normally pull, say the last 3 comments?

kalvdans replied:
I documented this in `GetCommentsReq` but maybe it is better to move all the documentation to the this rpc definition. I think a typical client will request 100 comments at a time.

aapeliv reviewed (approved):
This looks good, thanks! See a few small comments.

aapeliv commented:
Also, maybe this should be renamed to `threads` instead of `communities`?

kalvdans commented:
> Also, maybe this should be renamed to `threads` instead of `communities`?

The API namespace you mean? The thread is useless without a way to get ahold of a `thread_id`, which needs to be provided with other community-related calls. I thought all of them could be clumped under the communities namespace. But I can change, no problems.

kalvdans committed:
Update app/pb/communities.proto

Co-authored-by: amdwit <marco@witinformatisering.nl>

kalvdans committed:
Update app/pb/communities.proto

Co-authored-by: amdwit <marco@witinformatisering.nl>

Commit:
Pass 0 as out-of-band value

lucaslcode commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,105 @@
+  int64 comment_id = 1;
+  string content = 2;  // markdown without images
+  int64 author_user_id = 3;
+  google.protobuf.Timestamp created_time = 4;
```
Can this have a `num_replies` field?

lucaslcode reviewed (commented):
Do we want (now or later):
- up/downvotes?
- custom comment/reply sorting?
- Editing/deleting?
- Reporting or flagging?

aapeliv commented:
> Do we want (now or later):
> 
>     * up/downvotes?
> 
>     * custom comment/reply sorting?
> 
>     * Editing/deleting?
> 
>     * Reporting or flagging?

@ItsiW? What's the thinking with this?

lucaslcode commented:
I also think maybe there should be a way to get an overview of threads and replies so we don't have to choose between 20 gRPC calls and not loading any replies until asked. 

Maybe a `getThreadOverview` call which smartly returns the latest X comments along with the latest Y replies. If each comment includes a `num_replies` field frontend will be able to display a "load more" button

Also, I wonder if oldest first is a better sort order for replies.

Commit:
Duplicate some documentation in the rpc definition

Commit:
Add reply count to Comment

kalvdans commented:
> I also think maybe there should be a way to get an overview of threads and replies so we don't have to choose between 20 gRPC calls and not loading any replies until asked.
> Maybe a `getThreadOverview` call which smartly returns the latest X comments along with the latest Y replies. If each comment includes a `num_replies` field frontend will be able to display a "load more" button

I'm thinking getThreadOverview can be implemented by the fronted with one call to GetComments(), and then queueing up those 20 calls to GetReplies() in one HTTP stream, in the latency of two roundtrips to the server. Is it possible in the javascript frontend?

> Also, I wonder if oldest first is a better sort order for replies.

Not if smart clients want to show latest Y replies like you said :)

lucaslcode commented:
> Not if smart clients want to show latest Y replies

First Y replies, then!

> then queueing up those 20 calls to GetReplies() in one HTTP stream

Perhaps, but possibly not until we implement streaming rpc calls? Not so sure. But in any case, I think it would be a worse user experience - UI would jump around as replies are loaded and push down lower comments.

If you feel strongly that it should be managed in the frontend, how about allowing changing `getReplies` to accept [`oneof`](https://developers.google.com/protocol-buffers/docs/proto3#oneof) `comment_id` and `thread_id`? Where thread id just gives replies in the thread without worrying about the comments they are replying to.

Commit:
Use the same call regardless of level of the comment.

kalvdans commented:
I've unified the three levels (threads, comments, replies) into a single tree node type now. Still only latest comment first sorting available. I guess we can start here and implement earliest-first and depth-first later on.

aapeliv commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,81 @@
+
+  // Get non-deleted posts belonging to a thread, sorted in reverse time order.
+  // Pagination is supported by the posted_before_thread_id and number fields.
+  rpc ReadThread(ReadThreadReq) returns (ReadThreadRes);
```
Why `ReadThread` and not `GetThread`? I think `Get*` is more consistent and better due to its ubiquity in HTTP, etc.

aapeliv commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,81 @@
+  uint32 num_earlier_comments = 2;
+}
+
+message GetRepliesRes {
+  // sorted in inverse time order, latest reply first.
+  repeated Comment replies = 1;
+
+  // number of replies posted earlier than the ones returned in `replies`.
+  // zero if all replies has been returned to the client.
+  uint32 num_earlier_replies = 2;
+}
```
Is this leftover and unused?

aapeliv commented on app/pb/communities.proto:
```diff
@@ -0,0 +1,81 @@
+  // number of comments posted earlier than the ones returned in `comments`
+  // zero if all comments has been returned to the client.
+  uint32 num_earlier_comments = 2;
+}
```
Would it be good for this to return a "page token", i.e. the value to send to `posted_before_thread_id` to get an earlier list of messages? (I know this can be derived from `comments`). I'm not sure, what do you think?

kalvdans replied:
Yes, you pass the thread_id of the last item you got in the previous call :)

aapeliv replied:
That works!

aapeliv reviewed (approved):
This is looking very good, see a few small comments.

Is `Reply` a more descriptive word than `Comment` in this case? Up to you!

kalvdans committed:
Change read to get to be more http-alike

kalvdans committed:
Remove unused message

kalvdans committed:
Use pagination standard from https://google.aip.dev/158

kalvdans committed:
Only use thread and reply in the API, comment is confusing

kalvdans commented:
> Is `Reply` a more descriptive word than `Comment` in this case? Up to you!

It is a bit messy since a text could be both a comment or reply. But I also think Reply is better so I changed to that. I also switched to google's pagination. Now onto the implementation!

kalvdans committed:
First draft of implementation + test

kalvdans committed:
Add author to database

kalvdans committed:
pagination test and implementation

kalvdans committed:
rename communities -> threads

kalvdans committed:
black && isort && autoflake

kalvdans commented:
I've now implemented the backend side. The database still have separate tables for thread, comment and reply, even if the API has unified them. I've named everything "threads" to be easier to merge with the other communities stuff.

Missing:
- [x] database migrations, should I do them or disable the migration tests for now?
- [x] adding Threads servicer to the main app

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,103 @@
+class Threads(threads_pb2_grpc.ThreadsServicer):
+    def GetThread(self, request, context):
+        database_id, depth = unpack_thread_id(request.thread_id)
+        page_size = request.page_size or 1000
```
`PAGINATION_LENGTH`?

kalvdans replied:
Ah yes, thank you. Removed the unused definition ;)

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,103 @@
+            else:
+                context.abort(grpc.StatusCode.INVALID_ARGUMENT, "bad depth")
+            session.add(object_to_add)
+            session.flush()
+            session.refresh(object_to_add)
```
The `refresh` is superfluous. As long as you're still within the session ctx manager, it'll automatically refresh.

The `flush` is actually also unneeded. The transaction may fail upon leaving the context, which would happen after the `return` but before execution is passed back to whatever calls `PostReply`, so it will still fail with some error (I think `GrpcStatus.INTERNAL`).

I don't mind the `flush` though, as it's maybe more clear.

kalvdans replied:
The `flush()` seems to be needed to fill in the `id` member. It worked without the `refresh`, so will remove that!

aapeliv replied:
Ah yes, you're right. I missed that `.id`

aapeliv commented on app/backend/src/tests/test_fixtures.py:
```diff
@@ -277,13 +277,17 @@ def real_jail_session(token):
             server.stop(None).wait()
 
 
+def fake_channel(token):
+    user_id, jailed = Auth().get_session_for_token(token)
+    return FakeChannel(user_id=user_id)
+
+
```
Thanks, this is good. Could you also change the last remaining `api_session` to use `fake_channel`?

aapeliv commented on app/backend/src/tests/test_threads.py:
```diff
@@ -0,0 +1,125 @@
+    pass
+
+
+@contextlib.contextmanager
+def threads_session(token):
+    """
+    Create a communities API for testing, uses the token for auth
+    """
+    channel = fake_channel(token)
+    threads_pb2_grpc.add_ThreadsServicer_to_server(Threads(), channel)
+    yield threads_pb2_grpc.ThreadsStub(channel)
```
Should this go into `text_fixtures` like all other `*_session`s? Sometimes tests from other parts still need/want to call the APIs from another session. These are also fairly similar, so it's easier to refactor (as you've done in this PR with `fake_channel`) if they're all in the one place.

kalvdans replied:
No, let's move it later on when it is used by other test files. Regarding refactoring, it is as easy to refactor no matter how many files you touch (same amount of lines to be changed)

aapeliv commented on app/backend/src/tests/test_threads.py:
```diff
@@ -0,0 +1,125 @@
+    user1, token1 = generate_user()
+    user2, token2 = generate_user()
+
+    PARENT_THREAD_ID = 10
+
+    # Create a dummy Thread (should be replaced by pages later on)
+    with session_scope() as session:
+        session.add(Thread(id=1, title="foo"))
+
```
How about this?

```py
thread = Thread(id=1, title="foo")
session.add(thread)
session.flush()
PARENT_THREAD_ID = thread.id
```

I try to shy away from implementation details and execution details in tests and rely only on the public API. (i.e. the thread number is impl detail, and that it ends up having ID 10 is also a detail of execution)

kalvdans replied:
Thanks for the tip - applied

aapeliv commented on app/backend/src/tests/test_threads.py:
```diff
@@ -0,0 +1,125 @@
+
+
+def pagination_test(api, parent_id):
+    # Post some data
+    for c in string.ascii_lowercase:
+        api.PostReply(threads_pb2.PostReplyReq(thread_id=parent_id, content=c))
```
I prefer to use `random_hex` for random strings. Up to you!

kalvdans replied:
better keep the tests deterministic.

aapeliv commented on app/backend/src/tests/test_threads.py:
```diff
@@ -0,0 +1,125 @@
+    # Get it with pagination
+    token = ""
+    accumulator = []
+    while True:
+        ret = api.GetThread(threads_pb2.GetThreadReq(thread_id=parent_id, page_size=5, page_token=token))
+        assert 0 < len(ret.replies) <= 5
+        accumulator += [x.content for x in ret.replies]
+        assert len(accumulator) <= len(string.ascii_lowercase)
+        if len(accumulator) == len(string.ascii_lowercase):
+            assert ret.next_page_token == ""
+            break
+        assert ret.next_page_token != ""
+        assert len(ret.replies) == 5
+        token = ret.next_page_token
```
Might there be a case where this test fails and ends up with an infinite test that then screws up the test suite? Maybe `pytest` guards against this?

kalvdans replied:
It will not hang even on a buggy implementation, but I will switch to a for loop to make it clear!

aapeliv reviewed (changes requested):
See a few small comments.

aapeliv commented:
> I've now implemented the backend side. The database still have separate tables for thread, comment and reply, even if the API has unified them. I've named everything "threads" to be easier to merge with the other communities stuff.

Was it easier to implement this way and less hassle with trees? I would've imagined an arbitrarily-leveled single table would've been easier to implement.

But this looks good as well, and solid testing.

> database migrations, should I do them or disable the migration tests for now?
> adding Threads servicer to the main app

Yes, please add migrations (this is on top of `communities`, so we will recreate them before merging into `develop`), and please add the threads servicer into app as well.

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,103 @@
+PAGINATION_LENGTH = 10
+
+
+# Since the API exposes a single ID space regardless of nesting level,
+# we construct the API id by appending the the nesting level to the
+# database ID.
```
How about using a shared sequence instead? That would guarantee unique IDs across the two tables, and be less hassle from the backend?

kalvdans replied:
It will be more hassle in the backend as we can't see directly on the ID which table to select from.

kalvdans committed:
Remove refresh call as it is automatic.

kalvdans committed:
Remove unused constant

kalvdans committed:
Make use of the helper function just created

kalvdans committed:
Don't give a value to an autoincrement id

kalvdans committed:
Use a for loop to easily see that it terminates

kalvdans committed:
Merge remote-tracking branch 'origin/communities' into backend/feature/threads

kalvdans committed:
Disable migration tests for now, too lazy to add them

kalvdans committed:
black formatting

kalvdans committed:
Create a migration and enable migration tests again

kalvdans committed:
Add threads to servicer

kalvdans commented:
> Was it easier to implement this way and less hassle with trees? I would've imagined an arbitrarily-leveled single table would've been easier to implement.

I'm not comfortable with joins within the same table, but it can always be changed afterwards (but before we want permanent links)

> Yes, please add migrations (this is on top of `communities`, so we will recreate them before merging into `develop`), and please add the threads servicer into app as well.

Done!

aapeliv commented on app/backend/src/couchers/migrations/versions/152a14ce8d96_modify_the_database.py:
```diff
@@ -0,0 +1,48 @@
+    op.create_index(op.f("ix_discussions_thread_id"), "discussions", ["thread_id"], unique=True)
+    op.drop_index("ix_events_thread_id", table_name="events")
+    op.create_index(op.f("ix_events_thread_id"), "events", ["thread_id"], unique=True)
+    op.drop_index("idx_events_geom", table_name="events")
+    op.drop_index("idx_nodes_geom", table_name="nodes")
+    op.drop_index("idx_page_versions_geom", table_name="page_versions")
```
You should delete these.

I believe these are automatically generated indexes by PostGIS, that alembic unfortunately does not seem to realise. There may be a way to tweak alembic config, or somehow just add it to the models (though these need to be the right geospatial indexes).

kalvdans replied:
Bah, it is 2020, and still databases can't create their own indices based on which select queries that comes in.

aapeliv commented on app/backend/src/couchers/migrations/versions/152a14ce8d96_modify_the_database.py:
```diff
@@ -0,0 +1,48 @@
+    op.drop_column("replies", "author_user_id")
+    op.create_index("idx_page_versions_geom", "page_versions", ["geom"], unique=False)
+    op.create_index("idx_nodes_geom", "nodes", ["geom"], unique=False)
+    op.create_index("idx_events_geom", "events", ["geom"], unique=False)
```
Ditto.

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,99 @@
+
+
+def unpack_thread_id(thread_id):
+    return divmod(thread_id, 10)
```
Could you please add a comment as to document the return type of this function. I don't think it's obvious to everyone!

kalvdans replied:
black is putting my comments far away from the functions :/ but I've added a description of return value.

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,99 @@
+    return divmod(thread_id, 10)
+
+
+def pack_thread_id(database_id, depth):
+    return database_id * 10 + depth
```
Same (but not quite as important)

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,99 @@
+                ]
+
+            else:
+                context.abort(grpc.StatusCode.INVALID_ARGUMENT, "bad depth")
```
Maybe this could be `NOT_FOUND` (with error from `errors`).

That way the depth is just an implementation detail.

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,99 @@
+
+    def PostReply(self, request, context):
+        with session_scope() as session:
+            database_id, depth = unpack_thread_id(request.thread_id)
```
Not found?

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,99 @@
+            elif depth == 1:
+                object_to_add = Reply(comment_id=database_id, author_user_id=context.user_id, content=request.content)
+            else:
+                context.abort(grpc.StatusCode.INVALID_ARGUMENT, "bad depth")
```
Same thing about not found as `GetThread`?

aapeliv commented on app/pb/threads.proto:
```diff
@@ -0,0 +1,78 @@
+  // The same API is used regardless of level. All posts are using the same id namespace.
+
+  // Get non-deleted posts belonging to a thread, sorted in reverse time order.
+  // Pagination is supported by the posted_before_thread_id and number fields.
```
Stale `posted_before_thread_id`

aapeliv commented on app/pb/threads.proto:
```diff
@@ -0,0 +1,78 @@
+  int64 thread_id = 1;
+
+  // Maximum number of replies to return. Defaults to 1000.
+  int32 page_size = 2;
```
`uint32` as it's a count

kalvdans replied:
https://google.aip.dev/158 says `int32` but ok.

aapeliv reviewed (changes requested):
Looks mostly good. A few nitpicks! Some are not very important.

I would generally prefer descriptive variable names. Some of the ones in the generators (`r`, `n`, etc) are a bit confusing on first look.

aapeliv commented on app/backend/src/couchers/servicers/threads.py:
```diff
@@ -0,0 +1,99 @@
+                    .order_by(Comment.created.desc())
+                    .limit(page_size + 1)
+                    .all()
+                )
```
This doesn't seem to handle the case if the parent Comment doesn't exist. (Similarly below for Reply)

kalvdans replied:
Well spotted! Now fixed and tested.

kalvdans committed:
Document obscure return value

kalvdans committed:
Add tests for non-existing thread id:s

kalvdans committed:
black formatting

kalvdans committed:
Don't touch postgis internals!

kalvdans committed:
API documentation updates

kalvdans commented:
> I would generally prefer descriptive variable names. Some of the ones in the generators (`r`, `n`, etc) are a bit confusing on first look.

Thanks, I generally pick one-letter names when it is easy to grasp the scope. I'll think about longer names in future coding but I'm too lazy to update this branch now.

aapeliv reviewed (approved):
LGTM!
